src/Examples/CSharp DotNetCore/ExampleSimulator.cs

```diff
@@ -0,0 +1,190 @@
+                    await Task.Delay(500);
+                    Console.WriteLine($"{turnstileGate.Counter} people have passed through the gate");
+                }
+                
+            });
+
+            // The user input thread waits for the user to press "Enter", and then makes
+            // a request to recieve the latest data from the PLC.
+            var userInputThread = Task.Run(async () =>
+            {
+                Console.WriteLine("Press Enter to refresh the data from the remote device\n");
+
+                while (true)
+                {
+                    Console.ReadLine();
+                    await turnstileGate.ReceiveData();
+                }
+            });
+
+            Task.WaitAll(feedbackThread, userInputThread);
+        }
+
+
+        private interface ITurnstileGate
+        {
+            /// <summary>
+            /// Get the data from the PLC into HMI memory.
+            /// </summary>
+            Task ReceiveData();
+
+            /// <summary>
+            /// The most recent data recieved from the Turnstile Gate PLC regarding the counter.
+            /// </summary>
+            /// <remarks>
+            /// Note: This value will _not_ be automatically updated, instead we must call <see cref="ReceiveData"/> to update it.
+            /// </remarks>
+            int Counter { get; }
+
+            /// <summary>
+            /// Immediately sends a message to the Turnstile Gate PLC to reset the counter.
+            /// </summary>
+            Task ResetCounter();
+        }
+
+
+
+
+        /// <summary>
+        /// Provides a real implementation. Would be used when accessing the real device.
+        /// </summary>
+        private class RealImplementation : ITurnstileGate
+        {
+            private readonly Tag<DintPlcMapper, int> CounterTag;
+            private readonly Tag<BoolPlcMapper, bool> ResetTag;
+            public RealImplementation()
+            {
+                CounterTag = new Tag<DintPlcMapper, int>()
+                {
+                    PlcType = PlcType.ControlLogix,
+                    Path = "0,1",
+                    Name = "PEOPLE_COUNTER",
+                    Gateway = "10.10.10.10",
+                    Protocol = Protocol.ab_eip,
+                };
+
+                ResetTag = new Tag<BoolPlcMapper, bool>()
+                {
+                    PlcType = PlcType.ControlLogix,
+                    Path = "0,1",
+                    Name = "PEOPLE_COUNTER_RESET_FLAG",
+                    Gateway = "10.10.10.10",
+                    Protocol = Protocol.ab_eip
+                };
+            }
+
+            public int Counter { get; internal set; }
+
+            public async Task ResetCounter()
+            {
+                // The PLC is expected to have some logic that sets the ResetFlag from true back to false after resetting the counter
+                await ResetTag.WriteAsync(true);
+                await ReceiveData();
```

Comment:
I know this is just a throw away example, but it bothers me that this is a race condition...

Reply:
Well, in this case you're writing `true` to a PLC tag to presumably signal the PLC to reset a counter. And then you're reading the value of that counter with `ReceiveData`.  Depending how fast the PLC is scanning and how much network delay, etc you will either read back the value of the counter before the PLC processes the `ResetTag` trigger or you'll read it back after it processes `ResetTag`...

Like I said for the purpose of the example it's probably fine. In real life I'd probably set `ResetTag` to `true`, poll the tag until the PLC processes and sets it back to false (assuming PLC logic and order of operations), and then read the value of `CounterTag`.

Reply:
I see! Good points